Packages/ClientRuntime/Sources/Waiters/Waiter.swift

```diff
@@ -0,0 +1,79 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// An object used to wait on an operation until a desired state is reached, or until it is determined that
+/// the desired state will never be reached.
+/// Intended to be a generic type for use when waiting on any Smithy `operation`.
+/// May be reused for multiple waits, including concurrent operations.
+public class Waiter<Input, Output> {
+
+    /// The configuration this waiter was created with.
+    public let config: WaiterConfiguration<Input, Output>
+
+    /// The operation that this waiter will call one or more times while waiting on the success condition.
+    public let operation: (Input) async throws -> Output
+
+    /// Creates a `waiter` object with the supplied config and operation.
+    /// - Parameters:
+    ///   - config: An instance of `WaiterConfiguration` that defines the default behavior of this waiter.
+    ///   - operation: A closure that is called one or more times to perform the waiting operation;
+    ///   takes an `Input` as its sole param & returns an `Output` asynchronously.
+    ///   The `operation` closure throws an error if the operation cannot be performed or the
+    ///   operation completes with an error.
+    public init(
```

Comment:
For better DI, the Waiter shouldn't know how to create schedulers nor retryers, can we inject those details in? This would remove the need for duplicate `waitUntil` functions and would make `Waiter` a bit easier to test. If we don't want to expose this publicly, we can keep it internal.

```swift 
public init(
    config: WaiterConfiguration<Input, Output>,
    operation: @escaping (Input) async throws -> Output
) {
    self.config = config
    self.operation = operation
    self.init(
        config: config,
        operation: operation,
        scheduler: { options in
            let minDelay = options.minDelay ?? config.minDelay
            let maxDelay = options.maxDelay ?? config.maxDelay
            let maxWaitTime = options.maxWaitTime
            let scheduler = WaiterScheduler(minDelay: minDelay, maxDelay: maxDelay, maxWaitTime: maxWaitTime)
        },
        retryer: { WaiterRetryer<Input, Output>() }
    )
}

init(
    config: WaiterConfiguration<Input, Output>,
    operation: @escaping (Input) async throws -> Output,
    scheduler: (WaiterOptions) -> WaiterScheduler, 
    retryer: () -> WaiterRetryer<Input, Output>
) {
    self.config = config
    self.operation = operation
    self.scheduler = scheduler
    self.retryer = retryer
}
```

Reply:
Going to look at using factory pattern to create new schedulers & retryers, since a new retry & a new scheduler should be made for every request.

Reply:
Factory pattern works too, although it can be a bit heavy handed. This is a pseudo factory pattern tbh. Using closures should allow for new instances of scheduler and retryer to be created for each request